File: 01_Easy/0066_Plus_One.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {

        int input_size = digits.size();
        // create output vector
        vector<int> result(input_size);

        int carry_over = 1; // This is the "plusOne part"
        int digit;
        for (int i = input_size - 1; i >= 0; i--){
            // sum calculation
            digit = carry_over + digits.at(i);
            
            // calculate carry over and update the digit
            carry_over = digit / 10;
            digit = digit % 10;

            // store the digit in the output vector
            result.at(i) = digit;
        }

        // Add the carry over to the beginning of the number
        // Will always work in this simplistic way if the carry over is 9 or less; otherwise, we need to count the digits
        if (carry_over != 0){
            result.resize(input_size + 1);
            result.at(0) = carry_over;
        }

        return result;
    }
};

File: 01_Easy/0066_Plus_One.py
class Solution:
    def plusOne(self, digits: list[int]) -> list[int]:
        carry_over = 1
        idx = -1
        size = len(digits)
        while carry_over > 0 and abs(idx) <= size:
            if digits[idx] < 9:
                digits[idx] += 1
                carry_over = 0
            else:
                digits[idx] = 0
            idx -= 1
        
        if carry_over > 0:
            digits.insert(0, 1)
        
        return digits